### PR TITLE
854 persistent storage of dataset dependencies graph

### PR DESCRIPTION
**Pull request by zaychenko-sergei**
## Description

Closes: #854 

## Checklist before requesting a review

- [x] Unit and integration tests added
  
- [x] Compatibility:
    
  - [ ] Network APIs: ✅
    
  - [ ] Workspace layout and metadata: ✅
    
  - [ ] Configuration: ✅
    
  - [ ] Container images: ✅
- [ ] Observability:
    
  - [x] Tracing / [metrics](https://github.com/kamu-data/kamu-standards/blob/master/metrics_design.md): ✅
    
  - [ ] Alerts: ✅
- [ ] Documentation:
    
  - [x] [Changelog](./CHANGELOG.md): ✅
    
  - [ ] [Public documentation](https://github.com/kamu-data/kamu-docs/): ✅
  
- [ ] Downstream effects:
    
  - [x] [kamu-node](https://github.com/kamu-data/kamu-node): ✅
    
  - [ ] [kamu-web-ui](https://github.com/kamu-data/kamu-web-ui): ✅
    
  - [x] [release-queue](https://github.com/orgs/kamu-data/projects/4/views/1)

### CODE REVIEW

File: src/infra/datasets/postgres/src/repos/postgres_dataset_dependency_repository.rs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+#[async_trait::async_trait]
+impl DatasetDependencyRepository for PostgresDatasetDependencyRepository {
+    async fn stores_any_dependencies(&self) -> Result<bool, InternalError> {
+        let mut tr = self.transaction.lock().await;
+
+        let connection_mut = tr.connection_mut().await?;
+
+        let has_data = sqlx::query_scalar!(
+            r#"
+            SELECT EXISTS (SELECT * FROM dataset_dependencies LIMIT 1) as has_data
```
Comment (s373r) [inline src/infra/datasets/postgres/src/repos/postgres_dataset_dependency_repository.rs:44]: Q: out of curiosity, why not just `COUNT(*)`?

File: src/infra/datasets/repo-tests/src/dataset_dependencies_repository_test_suite.rs [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+                downstream_dataset_id: entry_baz.id,
+                upstream_dataset_ids: vec![entry_bar.id],
+            },
+        ]),
+    );
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub async fn test_dependency_fanins(catalog: &Catalog) {
```
Comment (s373r) [inline src/infra/datasets/repo-tests/src/dataset_dependencies_repository_test_suite.rs:171]:
```
test_dependency_fanins
                ^^^^^^
```
Q: out of curiosity, it's about https://en.wikipedia.org/wiki/Fan-in ?

File: src/domain/core/src/messages/core_message_types.rs [hunk 3]
```diff
@@ -49,6 +50,10 @@ impl DatasetLifecycleMessage {
         })
     }
 
+    pub fn about_2_delete(dataset_id: DatasetID) -> Self {
```
Comment (s373r) [inline src/domain/core/src/messages/core_message_types.rs:53]: Minor: `about_to_delete`?

**Review by s373r — COMMENTED** [review]
5 cents from my side

File: src/infra/datasets/postgres/src/repos/postgres_dataset_dependency_repository.rs (hunk 1 above)
Comment (zaychenko-sergei) [reply src/infra/datasets/postgres/src/repos/postgres_dataset_dependency_repository.rs:44]: I assume this query will at most look at 1 row, while `COUNT` is likely doing more work. Even if it's not loading all rows, it's still sort of a computation over table pages.

While we only need to know if table is empty...

File: src/domain/core/src/messages/core_message_types.rs (hunk 3 above)
Comment (zaychenko-sergei) [reply src/domain/core/src/messages/core_message_types.rs:53]: SMS slang from 90s.
to, two => 2
for => 4

File: src/infra/datasets/postgres/src/repos/postgres_dataset_dependency_repository.rs (hunk 1 above)
Comment (s373r) [reply src/infra/datasets/postgres/src/repos/postgres_dataset_dependency_repository.rs:44]: Yeah, makes a lot of sense -- thank you

File: src/domain/core/src/messages/core_message_types.rs (hunk 3 above)
Comment (s373r) [reply src/domain/core/src/messages/core_message_types.rs:53]: Yes, I understand and read it easily (btw like leetspeak: 1337, etc. just look at my nickname :) )

I mean,
Q: do we need to use it in the naming? I don't see that kind of slang anywhere else in the code base

File: src/domain/core/src/messages/core_message_types.rs [hunk 4]
```diff
@@ -23,8 +23,9 @@ const DATASET_LIFECYCLE_OUTBOX_VERSION: u32 = 1;
 pub enum DatasetLifecycleMessage {
     Created(DatasetLifecycleMessageCreated),
     DependenciesUpdated(DatasetLifecycleMessageDependenciesUpdated),
-    Deleted(DatasetLifecycleMessageDeleted),
     Renamed(DatasetLifecycleMessageRenamed),
+    About2Delete(DatasetLifecycleMessageAbout2Delete),
```
Comment (sergiimk) [inline src/domain/core/src/messages/core_message_types.rs:27]: Why do we need to separate "about to delete" and "deleted" messages?

I thought that one benefit of combining `immediate` and `durable` delivery mechanisms in one messaging system is that things that need to happen inside of the same transaction can subscribe to the same event as those that need to happen outside of the transaction.

Here it almost feels like `About2Delete` event is intended ONLY to be delivered via `immediate` mode, which seems strange as delivery is currently a property of a consumer.

File: src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+use internal_error::InternalError;
+use opendatafabric::DatasetID;
+use thiserror::Error;
+use tokio_stream::Stream;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait DatasetDependencyRepository: Send + Sync {
+    async fn stores_any_dependencies(&self) -> Result<bool, InternalError>;
```
Comment (sergiimk) [inline src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs:21]: How about replacing this with `edges_count()` as a more versatile method.

File: src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    async fn remove_all_dependencies_of(&self, dataset_id: &DatasetID)
+        -> Result<(), InternalError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct DatasetDependencies {
+    pub downstream_dataset_id: DatasetID,
+    pub upstream_dataset_ids: Vec<DatasetID>,
```
Comment (sergiimk) [inline src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs:46]: Why don't we stream just the ID pairs (as they are stored in the DB) and let the caller aggregate however they want? Seems like an unnecessary complication of a (what should be low-level) interface.

**Review by sergiimk — APPROVED** [review]
Left some suggestions and questions, but looks great overall :+1:

File: src/domain/core/src/messages/core_message_types.rs (hunk 3 above)
Comment (zaychenko-sergei) [reply src/domain/core/src/messages/core_message_types.rs:53]: Зануды!

File: src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs (hunk 5 above)
Comment (zaychenko-sergei) [reply src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs:21]: We only need to know if the repository is empty.
Is empty != Count edges from performance point of view.
If we will need to know the number of edges, we will add a separate API.

Somehow, standard containers have both `is_empty` and `len`.
Especially with more advanced structures like linked lists and trees this  this makes a lot more sense, right?
So, it's the same, I'm hoping to use cheaper queries for quick tests.

File: src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs (hunk 6 above)
Comment (zaychenko-sergei) [reply src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs:46]: We already had this modeling before. I don't see a motivation to reduce the level of abstraction.

File: src/domain/core/src/messages/core_message_types.rs (hunk 4 above)
Comment (zaychenko-sergei) [reply src/domain/core/src/messages/core_message_types.rs]: I've tried cascade deletions on foreign keys in new tables, and this worked better, so `AboutToDelete` is no longer needed.

File: src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs (hunk 6 above)
Comment (zaychenko-sergei) [reply src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs:46]: Inserts are looking quite nicely currently.

**Pull request merged**